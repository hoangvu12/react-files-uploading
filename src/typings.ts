import React from 'react';

export interface FileUploadingPropsType {
  value: File[];
  onChange: (value: File[], addUpdatedIndex?: Array<number>) => void;
  children?: (props: ExportInterface) => React.ReactNode;
  multiple?: boolean;
  maxNumber?: number;
  acceptType?: Array<string>;
  maxFileSize?: number;
  onError?: (errors: ErrorsType, files?: File[]) => void;
  inputProps?: React.HTMLProps<HTMLInputElement>;
}

export interface ExportInterface {
  fileList: File[];
  onFileUpload: () => void;
  onFileRemoveAll: () => void;
  errors: ErrorsType;
  onFileUpdate: (index: number) => void;
  onFileRemove: (index: number) => void;
  isDragging: boolean;
  dragProps: {
    onDrop: (e: any) => void;
    onDragEnter: (e: any) => void;
    onDragLeave: (e: any) => void;
    onDragOver: (e: any) => void;
    onDragStart: (e: any) => void;
  };
}

export type ErrorsType = {
  maxFileSize?: boolean;
  maxNumber?: boolean;
  acceptType?: boolean;
} | null;
