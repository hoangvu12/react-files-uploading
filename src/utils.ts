import React from 'react';

export const openFileDialog = (
  inputRef: React.RefObject<HTMLInputElement>
): void => {
  if (inputRef.current) inputRef.current.click();
};

export const getAcceptTypeString = (acceptType?: Array<string>) => {
  return acceptType && acceptType.length > 0
    ? acceptType.map((item) => `.${item}`).join(', ')
    : '*/*';
};
