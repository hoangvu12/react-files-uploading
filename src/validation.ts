import { DEFAULT_NULL_INDEX } from './constants';
import { ErrorsType } from './typings';

export const isMaxFileSizeValid = (fileSize: number, maxFileSize?: number) => {
  return maxFileSize ? fileSize <= maxFileSize : true;
};

export const isAcceptTypeValid = (
  acceptType: string[] | null,
  fileName: string
) => {
  if (acceptType && acceptType.length > 0) {
    const type: string = fileName.split('.').pop() || '';
    if (
      acceptType.findIndex(
        (item) => item.toLowerCase() === type.toLowerCase()
      ) < 0
    )
      return false;
  }
  return true;
};

export const isMaxNumberValid = (
  totalNumber: number,
  maxNumber: number | null,
  keyUpdate: number | null
) => {
  if (maxNumber !== 0 && !maxNumber) return true;

  if (keyUpdate === DEFAULT_NULL_INDEX) {
    if (totalNumber <= maxNumber) return true;
  } else if (totalNumber <= maxNumber + 1) return true;

  return false;
};

export const getErrorValidation = ({
  fileList,
  value,
  maxNumber,
  keyUpdate,
  acceptType,
  maxFileSize,
}): ErrorsType => {
  const newErrors: ErrorsType = {};

  if (!isMaxNumberValid(fileList.length + value.length, maxNumber, keyUpdate)) {
    newErrors.maxNumber = true;
  } else {
    for (let i = 0; i < fileList.length; i += 1) {
      const { file } = fileList[i];
      if (!file) continue;

      if (!isAcceptTypeValid(acceptType, file.name)) {
        newErrors.acceptType = true;
        break;
      }

      if (!isMaxFileSizeValid(file.size, maxFileSize)) {
        newErrors.maxFileSize = true;
        break;
      }
    }
  }

  if (Object.values(newErrors).find(Boolean)) return newErrors;

  return null;
};
