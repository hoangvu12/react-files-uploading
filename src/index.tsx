import React, { useRef, useState, useCallback, useMemo } from 'react';
import { openFileDialog, getAcceptTypeString } from './utils';
import { getErrorValidation } from './validation';
import { FileUploadingPropsType, ErrorsType } from './typings';
import { DEFAULT_NULL_INDEX, INIT_MAX_NUMBER } from './constants';

const ReactFileUploading: React.FC<FileUploadingPropsType> = ({
  value = [],
  onChange,
  onError,
  children,
  multiple = false,
  maxNumber = INIT_MAX_NUMBER,
  acceptType,
  maxFileSize,
  inputProps = {},
}) => {
  const inValue = value || [];
  const inputRef = useRef<HTMLInputElement>(null);
  const [keyUpdate, setKeyUpdate] = useState<number>(DEFAULT_NULL_INDEX);
  const [errors, setErrors] = useState<ErrorsType>(null);
  const [isDragging, setIsDragging] = useState<boolean>(false);

  const handleClickInput = useCallback(
    () => openFileDialog(inputRef),
    [inputRef]
  );

  const onFileUpload = useCallback((): void => {
    setKeyUpdate(DEFAULT_NULL_INDEX);
    handleClickInput();
  }, [handleClickInput]);

  const onFileRemoveAll = useCallback((): void => {
    onChange?.([]);
  }, [onChange]);

  const onFileRemove = (index: number | Array<number>): void => {
    const updatedList = [...inValue];
    if (Array.isArray(index)) {
      index.forEach((i) => {
        updatedList.splice(i, 1);
      });
    } else {
      updatedList.splice(index, 1);
    }
    onChange?.(updatedList);
  };

  const onFileUpdate = (index: number): void => {
    setKeyUpdate(index);
    handleClickInput();
  };

  const validate = async (fileList: File[]): Promise<boolean> => {
    const errorsValidation = getErrorValidation({
      fileList,
      maxFileSize,
      maxNumber,
      acceptType,
      keyUpdate,
      value: inValue,
    });

    if (errorsValidation) {
      setErrors(errorsValidation);
      onError?.(errorsValidation, fileList);
      return false;
    }

    errors && setErrors(null);

    return true;
  };

  const handleChange = async (files: FileList | null) => {
    if (!files?.length) return;

    const fileList = Array.from(files);

    const isValidate = await validate(fileList);

    if (!isValidate) return;

    let updatedFileList: File[] = [];

    const updatedIndexes: number[] = [];

    if (keyUpdate > DEFAULT_NULL_INDEX) {
      const firstFile = files[0];
      updatedFileList = [...inValue];
      updatedFileList[keyUpdate] = firstFile;
      updatedIndexes.push(keyUpdate);
    } else if (multiple) {
      updatedFileList = [...inValue, ...fileList];
      for (
        let i = inValue.length as number;
        i < updatedFileList.length;
        i += 1
      ) {
        updatedIndexes.push(i);
      }
    } else {
      updatedFileList = [fileList[0]];
      updatedIndexes.push(0);
    }

    onChange?.(updatedFileList, updatedIndexes);
  };

  const onInputChange = async (
    e: React.ChangeEvent<HTMLInputElement>
  ): Promise<void> => {
    await handleChange(e.target.files);
    keyUpdate > DEFAULT_NULL_INDEX && setKeyUpdate(DEFAULT_NULL_INDEX);
    if (inputRef.current) inputRef.current.value = '';
  };

  const acceptTypeString = useMemo(
    () => getAcceptTypeString(acceptType),
    [acceptType]
  );

  const handleDrag = (e: React.DragEvent<HTMLDivElement>) => {
    e.preventDefault();
    e.stopPropagation();
  };

  const handleDragIn = (e: React.DragEvent<HTMLDivElement>) => {
    e.preventDefault();
    e.stopPropagation();
    if (e.dataTransfer.items && e.dataTransfer.items.length > 0) {
      setIsDragging(true);
    }
  };

  const handleDragOut = (e: React.DragEvent<HTMLDivElement>) => {
    e.preventDefault();
    e.stopPropagation();
    setIsDragging(false);
  };

  const handleDrop = (e: React.DragEvent<HTMLDivElement>) => {
    e.preventDefault();
    e.stopPropagation();
    setIsDragging(false);
    if (e.dataTransfer.files && e.dataTransfer.files.length > 0) {
      handleChange(e.dataTransfer.files);
    }
  };

  const handleDragStart = (e: React.DragEvent<HTMLDivElement>) => {
    e.preventDefault();
    e.stopPropagation();
    e.dataTransfer.clearData();
  };

  return (
    <>
      <input
        type="file"
        accept={acceptTypeString}
        ref={inputRef}
        multiple={multiple && keyUpdate === DEFAULT_NULL_INDEX}
        onChange={onInputChange}
        style={{ display: 'none' }}
        {...inputProps}
      />
      {children?.({
        fileList: inValue,
        onFileUpload,
        onFileRemoveAll,
        onFileUpdate,
        onFileRemove,
        errors,
        dragProps: {
          onDrop: handleDrop,
          onDragEnter: handleDragIn,
          onDragLeave: handleDragOut,
          onDragOver: handleDrag,
          onDragStart: handleDragStart,
        },
        isDragging,
      })}
    </>
  );
};

export default ReactFileUploading;

export { FileUploadingPropsType, ErrorsType };
