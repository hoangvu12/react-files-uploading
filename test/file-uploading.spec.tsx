import * as React from 'react';
import sinon from 'sinon';
import { render, fireEvent, waitFor } from '@testing-library/react';
import App from './fixtures/app';
import * as Utils from '../src/utils';

describe('testing ReactImageUploading component', () => {
  let sandbox;
  let openDialog;

  beforeEach(() => {
    sandbox = sinon.createSandbox();
    openDialog = sandbox.stub(Utils, 'openFileDialog');
  });

  afterEach(() => {
    sandbox.restore();
  });

  it('should open dialog', () => {
    const { getByText } = render(<App />);
    fireEvent.click(getByText(/Click or Drop here/));
    expect(openDialog.called).toBeTruthy();
  });

  it('should upload a file', async () => {
    const { queryByTestId, getByText } = render(<App />);

    // 1. Open file dialog
    fireEvent.click(getByText(/Click or Drop here/));
    expect(openDialog.called).toBeTruthy();

    // 2. Mock file data
    const file = new File([''], 'test_video.mp4', { type: 'video/mp4' });
    const inputEl = document.querySelectorAll(
      '[type="file"]'
    )[0] as HTMLInputElement;

    Object.defineProperty(inputEl, 'files', { value: [file] });

    fireEvent.change(inputEl);

    await waitFor(() => queryByTestId('file-0'));

    const fileName = (queryByTestId('file-0') as HTMLParagraphElement)
      .innerText;
    expect(fileName).toMatchSnapshot('test_video');
  });

  it('should upload multiple files', async () => {
    const { queryByTestId, getByText } = render(<App multiple />);

    // 1. Open file dialog
    fireEvent.click(getByText(/Click or Drop here/));
    expect(openDialog.called).toBeTruthy();

    // 2. Mock file data
    const videoFile = new File([''], 'test_video.mp4', { type: 'video/mp4' });
    const imageFile = new File([''], 'test_image.mp4', { type: 'image/png' });
    const textFile = new File([''], 'test_text.txt', { type: 'text/plain' });

    const inputEl = document.querySelectorAll(
      '[type="file"]'
    )[0] as HTMLInputElement;

    Object.defineProperty(inputEl, 'files', {
      value: [videoFile, imageFile, textFile],
    });

    fireEvent.change(inputEl);

    await waitFor(() => queryByTestId('file-0'));

    const videoFileName = (queryByTestId('file-0') as HTMLParagraphElement)
      .innerText;
    const imageFileName = (queryByTestId('file-1') as HTMLParagraphElement)
      .innerText;
    const textFileName = (queryByTestId('file-2') as HTMLParagraphElement)
      .innerText;

    expect(videoFileName).toMatchSnapshot('test_video');
    expect(imageFileName).toMatchSnapshot('test_image');
    expect(textFileName).toMatchSnapshot('test_text');
  });

  it('should remove all videos', async () => {
    const { queryByTestId, getByText } = render(<App />);

    fireEvent.click(getByText(/Click or Drop here/));
    expect(openDialog.called).toBeTruthy();

    const file = new File([''], 'test_video.mp4', { type: 'video/mp4' });
    const inputEl = document.querySelectorAll(
      '[type="file"]'
    )[0] as HTMLInputElement;

    Object.defineProperty(inputEl, 'files', { value: [file] });

    fireEvent.change(inputEl);

    await waitFor(() => queryByTestId('file-0'));

    fireEvent.click(getByText(/Remove all videos/));
    expect(queryByTestId('file-0')).toBeNull();
  });

  it('should remove specific video', async () => {
    const { queryByTestId, getByText } = render(<App />);

    fireEvent.click(getByText(/Click or Drop here/));
    expect(openDialog.called).toBeTruthy();

    const file = new File([''], 'test_video.mp4', { type: 'video/mp4' });
    const inputEl = document.querySelectorAll(
      '[type="file"]'
    )[0] as HTMLInputElement;

    Object.defineProperty(inputEl, 'files', { value: [file] });

    fireEvent.change(inputEl);

    await waitFor(() => queryByTestId('file-0'));

    fireEvent.click(getByText('Remove 0'));
    expect(queryByTestId('file-0')).toBeNull();
  });
});
