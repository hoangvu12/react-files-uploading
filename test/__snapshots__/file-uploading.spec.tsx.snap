// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing ReactImageUploading component should upload a file: test_video 1`] = `undefined`;

exports[`testing ReactImageUploading component should upload multiple files: test_image 1`] = `undefined`;

exports[`testing ReactImageUploading component should upload multiple files: test_text 1`] = `undefined`;

exports[`testing ReactImageUploading component should upload multiple files: test_video 1`] = `undefined`;
